src/Analyzers/CSharp/CodeFixes/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.cs

```diff
@@ -82,10 +82,27 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                 }
 
                 var switchStatement = (SwitchStatementSyntax)switchLocation.FindNode(getInnermostNodeForTie: true, cancellationToken);
+
+                // If the swith statement expression is being implicitly converted then we need to explicitly cast the expression
+                // before rewriting as a switch expression
+                var expressionISymbolType = semanticModel.GetSymbolInfo(switchStatement.Expression).Symbol.GetSymbolType();
+                var expConversionISymbolType = semanticModel.GetTypeInfo(switchStatement.Expression).ConvertedType;
+
+                var needsCast = false;
+                SwitchStatementSyntax newSwitch = null;
+                if (expConversionISymbolType != expressionISymbolType && expConversionISymbolType != null)
+                {
+                    needsCast = true;
```

Comment:
instead of this variable and the need to use a ternary operator why not add a local function

```csharp
static SwitchStatementSyntax AddCastIfNecessary(SemanticModel model, SwitchStatementSyntax switchStatement)
{
    // If the swith statement expression is being implicitly converted then we need to explicitly cast the expression
    // before rewriting as a switch expression
    var expressionISymbolType = model.GetSymbolInfo(switchStatement.Expression).Symbol.GetSymbolType();
    var expConversionISymbolType = model.GetTypeInfo(switchStatement.Expression).ConvertedType;

    if (expConversionISymbolType != null &&
        !SymbolEqualityComparer.Default.Equals(expConversionISymbolType, expressionISymbolType))
    {
        return switchStatement.Update(switchStatement.SwitchKeyword, switchStatement.OpenParenToken,
            switchStatement.Expression.Cast(expConversionISymbolType).WithAdditionalAnnotations(Formatter.Annotation),
            switchStatement.CloseParenToken, switchStatement.OpenBraceToken,
            switchStatement.Sections, switchStatement.CloseBraceToken);
    }

    return switchStatement;
}
```

then the code just becomes
```csharp
switchStatement = AddCastIfNecessary(semanticModel, switchStatement);

var switchExpression = Rewriter.Rewrite(switchStatement,
                                        declaratorToRemoveTypeOpt,
                                        nodeToGenerate,
                                        shouldMoveNextStatementToSwitchExpression: shouldRemoveNextStatement,
                                        generateDeclaration: declaratorToRemoveLocationOpt is object);
```

Reply:
I moved the cast into the ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter and it now casts every time. The cast is removed by the Simplifier if it is not needed.